source/opt/instruction.cpp

```diff
@@ -389,5 +395,49 @@ Instruction* Instruction::InsertBefore(std::unique_ptr<Instruction>&& i) {
   i.get()->InsertBefore(this);
   return i.release();
 }
+
+bool Instruction::IsValidBasePointer() const {
+  uint32_t tid = type_id();
+  if (tid == 0) {
+    return false;
+  }
+
+  ir::Instruction* type = context()->get_def_use_mgr()->GetDef(tid);
+  if (type->opcode() != SpvOpTypePointer) {
+    return false;
+  }
+
+  if (context()->module()->HasCapability(SpvCapabilityAddresses)) {
+    // TODO: The rules here could be more restrictive.
+    return true;
+  }
+
+  if (opcode() == SpvOpVariable || opcode() == SpvOpFunctionParameter) {
+    return true;
+  }
+
+  uint32_t pointee_type_id = type->GetSingleWordInOperand(1);
+  ir::Instruction* pointee_type_inst =
+      context()->get_def_use_mgr()->GetDef(pointee_type_id);
+
+  if (pointee_type_inst->IsOpaqueType()) {
+    return true;
+  }
+  return false;
+}
+
+bool Instruction::IsOpaqueType() const {
+  if (opcode() == SpvOpTypeStruct) {
```

Comment:
I think you need to check the element type of arrays too. You could have an array of opaque types I believe. I see no restriction in the spec.